### ISSUE TITLE
Interactive prompt doesn't properly work with arrow and backspace keys in haiku terminal

### ISSUE THREAD

**Issue by FractalU** [haikuports/haikuports#8972]
When I open either lua or python interactive prompt I notice that arrow and backspace keys don't properly work. When pressing an arrow key, instead of moving the cursor or navigating through the command history a big letter is being printed. Below you can see which letter is being printed from the arrow keys.
```
Arrow Up -> A
Arrow Right -> C
Arrow Down -> B
Arrow Left -> D
```
The backspace key also doesn't properly work, but it's different. If pressing the backspace key too many times. Instead of stopping at the beginning, it deletes the printed prompt characters as well and the cursor is too far behind.

It turns out that it has something to do with reading from the `stdin` stream in C/C++. I've written a sample C program to reproduce this issue.

Here is the C code.
```
#include <stdio.h>
#include <string.h>

void trim_eol_inline(char *s) {
	int n = strlen(s);
	if (n > 0) {
		for (int i = n - 1; s[i] == '\n' || s[i] == '\r'; i--) {
			s[i] = '\0';
		}
	}
}

int main(int argc, char **argv) {
	char sbuf[100];
	
	printf("input: ");
	if (!fgets(sbuf, sizeof(sbuf) / sizeof(sbuf[0]), stdin)) {
		return 1;
	}
	trim_eol_inline(sbuf);
	
	printf("text: '%s'\n", sbuf);
	printf("number of characters: %d\n", strlen(sbuf));
	
	return 0;
}
```
Here I use the `fgets` function to read from the `stdin` stream.

As a test, type all the arrow keys as `↑→↓←` in the prompt. The output should look as follows.
```
input: ACBD
text: 'ACBD'
number of characters: 12
```
Notice that each arrow key takes up three characters.

To test backspace type the backspace key three times follow by 'a'. The output should look as follows.
```
inpua
text: 'a'
number of characters: 1
```
Notice how the backspace key deletes the printed prompt characters and that the cursor is too far behind.

**Comment by pulkomandy** [comment]
Your test program shows the normal behavior, you can run it on other systems (Linux for example) and get similar results. The arrow keys send specific escape sequences that are normally handled by readline or ncurses (in the case of Python this should be readline).

**Comment by FractalU** [comment]
That's interesting.
Maybe I should be more specific. I came across this issue in haiku terminal. I use haiku `R1/beta4 (Revision hrev56578+87)` in virtualbox. I hope that helps. What I experienced in haiku terminal is not the normal behavior.

**Comment by augiedoggie** [comment]
I haven't looked into it but there seems to be something different about the Haiku Terminal and arrow keys. I have to have different mappings in my zsh config for `Ctrl-<left>`/`Ctrl-<right>` compared to Linux.  Also, I can see differences with the regular arrow keys as well when I use the verbatim command(`Ctrl-v`) in a shell.  For example, `Ctrl-v <left>` gives me `^[[D` on Linux but `^[OD` on Haiku.

**Comment by pulkomandy** [comment]
Yes, the escape sequences are different, but the terminfo database will take care of that. See the discussion in https://review.haiku-os.org/c/haiku/+/84 for more details.

**Comment by FractalU** [comment]
Thanks for the link. This partially explains why when I press the arrow keys it shows the letters which are three characters wide. From the `VTkeymap.h` file I can look up the actual key codes.

Here are the key code definitions in `src/apps/terminal/VTkeymap.h`.
```
#define LEFT_ARROW_KEY_CODE "\033OD"
#define RIGHT_ARROW_KEY_CODE "\033OC"
#define UP_ARROW_KEY_CODE "\033OA"
#define DOWN_ARROW_KEY_CODE "\033OB"

#define CTRL_LEFT_ARROW_KEY_CODE "\033O5D"
#define CTRL_RIGHT_ARROW_KEY_CODE "\033O5C"
#define CTRL_UP_ARROW_KEY_CODE "\033O5A"
#define CTRL_DOWN_ARROW_KEY_CODE "\033O5B"
```
Apparently, the interactive prompt in haiku terminal doesn't handle the key codes correctly. Instead, it prints the last printable characters from the key codes and stores those key codes into the input buffer. Here is another example with pressing control and arrow keys together.

When pressing `(Ctrl + ↑)(Ctrl + →)(Ctrl + ↓)(Ctrl + ←)` I get the following output.
```
input: 5A5C5B5D
text: '5A5C5B5D'
number of characters: 16
```
Now, those keys are four characters wide. This is expected from the key code definitions. I get similar behavior with the insert, delete, home, end, page up, page down, scroll lock, pause/break, escape, function keys. The control, alt, shift, caps lock keys work normally. The backspace key kind of works. Just the cursor can go too far back.

I hope any of you could reproduce this issue in the haiku terminal.

**Comment by augiedoggie** [comment]
> Yes, the escape sequences are different, but the terminfo database will take care of that. See the discussion in https://review.haiku-os.org/c/haiku/+/84 for more details.

Yes, but it appears our lua is built without readline/ncurses support.

**Comment by augiedoggie** [comment]
Arrow keys work fine in the `lua` repl after enabling readline.  I'll see about pushing a patch later today if possible.

**Comment by FractalU** [comment]
> Arrow keys work fine in the lua repl after enabling readline. I'll see about pushing a patch later today if possible.

Wait a moment. Lua repl typically uses `fgets` to read from the interactive prompt. This is the same mechanism as my test program above. What I expect is that all the C functions that read from `stdio` such as `fgets`, `gets`, `getchar`, etc. properly work. If you replace those C function with readline on some programs such as lua, those programs work but the other ones don't. That's a horrible solution. Also, you patch a lot of programs which would run perfectly fine if there is a system wide solution. Think through this again carefully.

**Comment by augiedoggie** [comment]
Almost all of the other systems use libreadline/libedit in lua.  I'm not hacking in some functionality, I'm just enabling the switch in the Makefile.

**Comment by augiedoggie** [comment]
Which version of `python` are you executing?  The regular `python` command on my system is python 2, which appears to be built without line editing support(I haven't looked at the recipe).  However, when I execute `python3.10` then the REPL functions properly.

**Comment by FractalU** [comment]
I use python 2.7.18-2 which is the regular python. The regular python has the same buggy behavior with the REPL though.

For the lua case I do agree enabling `readline` in the makefile is the best solution right now. But in the long term I would like to have the C function which read from `stdin` fixed. Though this requires changes in `libc` which is not gonna be simple. I do know that on linux it's similarly buggy with those C functions, but it's better on the linux terminal than on the haiku terminal, because on the linux terminal it prints the whole key code whereas on the haiku terminal it only prints the last printable characters such as letters or digits, which can be far more confusing. Additionally, the backspace key on the haiku terminal can set the cursor too far back, which doesn't happen on the linux terminal.

**Comment by pulkomandy** [comment]
Python 2 was modified to use libedit but I think the Python buildsystem doesn't support this and only works with libreadline (they are generally API compatible but there are some differences). It could be switched again, but really people should be using Python 3 now, Python 2 is dead.

**Comment by pulkomandy** [comment]
Lua was fixed and Pyhton is now by default the 3.10 version which works correctly. I think nothing more is left to be done here.

**Comment by FractalU** [comment]
Lua was fixed using `readline`? What about the stdio C functions which read from the `stdin` stream such as `fgets`, `fgetc`, `getc`, `getchar`, etc?

**Comment by pulkomandy** [comment]
These behave as they do on all other systems where they also don't handle this directly. There is nothing to fix for them.

**Comment by FractalU** [comment]
I do agree that they behave the same on all other unix systems. However, on those systems they still do a better job with the prompt. They print the whole control characters as opposed to last printable characters and the backspace fully works. That's not the case with the haiku terminal. I already explained this in more detailed previously. In addition, luajit doesn't have `readline` as opposed to `lua`. That means `lua` got fixed but not `luajit`. The `readline` api is a workaround solution. where it fixed some programs, but not the others.
